_posts/2017-10-27-conventional-interfaces.md

```diff
@@ -0,0 +1,225 @@
+* The *folding* or reduction is partially implemented by the `+` that joins the double recursion in the `Node` pattern match, and partially by the `otherwise` case in the `Leaf` pattern match.
+
+Each pattern matching is mixing several concerns, increasing the complexity of the implementation and hiding the signal-flow structure of the computation.
+
+`evenFibs`:
+
+* The *enumeration* is partially implemented by the `k > n` case and party by the `next` expression.
+* The *mapping*, although extracted to the `f` expression, is mixed with the `even` filtering in the `otherwise` case.
+* The *folding* or reduction is partially implemented by `:`, the list constructor, and partially by the `[ ]`, empty list, in the `k > n` case.
+
+As in `sumOddSquares`, each pattern matching in `evenFibs` is mixing different concerns, failing to exhibit the signal-flow structure of the computation.
+
+Let's refactor both functions following the steps defined previously, so that we can expose the [hidden similiarities](#similarities) and the signal-flow structure.
+
+It is worth noting that `.` is the operator for function composition in Haskell.
+
+`sumOddSquares`:
+
+1. *Enumerates* the leaves of the input tree.
+2. *Filters* out the leaves that contain even numbers.
+3. *Maps* the remaining leaves squaring each one.
+4. *Folds* (or reduces) the mapped leaves using `+`, starting at 0.
+
+```
+sumOddSquares :: (Integral a) => BinaryTree a -> a
+sumOddSquares tree = foldr (+) 0 . fmap (^2) . filter odd (enumerateLeaves tree)
+
+enumerateLeaves :: BinaryTree a -> [a]
+enumerateLeaves tree = foldr (:) [] tree
+```
+
+`evenFibs`:
+
+1. *Enumerates* the interval 0 to n.
+2. *Maps* each number to its corresponding Fib.
+3. *Filter* out the odd Fibs.
+4. *Folds* (or reduces) the remaining elements using `:` (List constructor), starting with `[]` (empty list).
+
+```
+evenFibs' :: Integer -> [Integer]
+evenFibs' to = foldr (:) [] . filter even . fmap fib (enumerateInterval 0 to)
+
+enumerateInterval :: Integer  -> Integer -> [Integer]
+enumerateInterval from to = [from..to]
+```
+## Signal processing metaphor
+
+Now that we have seen how different computations can be shaped similarly by using the same set of combinators composed in different ways, let's expand on the signal processing metaphor.
+
+As mentioned in SICP, the key to organizing programs so as to more clearly reflect the signal-flow structure is to concentrate on the "signals" that flow from one stage in the process to the next. SCIP uses Lisp, where everything is a list, therefore the signal processing metaphor fits very nicely with lists. If we move away from Lisp and to Haskell for example, the metaphor may not be as evident, but as we just showed it is equally applicable. In most of the cases, Haskell abstracts the different combinators from the concrete data structure using [type classes](https://en.wikipedia.org/wiki/Type_class), let's break down a few of them.
```

Comment:
"from Lisp and to Hakell"
substitute for
"from  Lisp to Haskell" no need of and

Reply:
Done